crates/analyzer-core/src/lib.rs

```diff
@@ -187,14 +216,27 @@ pub fn lex(db: &dyn crate::Db, file_id: FileId, buf: Buffer) -> LexedBuffer {
 }
 
 #[salsa::tracked(return_ref)]
-pub fn preprocess(db: &dyn crate::Db, fs: Fs, file_id: FileId) -> Option<Vec<(FileId, Token, Span)>> {
+pub fn preprocess<'a>(db: &dyn crate::Db, fs: Fs, file_id: FileId) -> Option<Vec<(FileId, Token, Span)>> {
 	let mut pp = PreprocessorState::new(
-		|path: String| FileId::new(db, path),
+		|path: &str| {
+			// Return a `FileId` with an absolute path resolved relative to the file being preprocessed. If the path
+			// is already absolute then `resolve_path` will return it as is.
+			let base_url = file_id.path(db);
+			let target_path = db.resolve_path(&base_url, path);
+
+			FileId::new(db, target_path)
+		},
 		|file_id| {
-			fs.fs(db).get(&file_id).map(|&buf| {
-				let lexed = lex(db, file_id, buf);
-				lexed.lexemes(db)
-			})
+			let path = file_id.path(db); // Get the path from the `FileId` and ensure its stored for the dependency.
+
+			match fs.fs(db).get(&file_id) {
+				Some(buf) => Some(lex(db, file_id, *buf).lexemes(db)),
+				None => {
+					IncludedDependencies::push(db, IncludedDependency { file: file_id, include_path: path.into() });
+
+					None
+				}
+			}
```

Comment:
Both branches of the `match` should push to `IncludedDependencies`. `IncludedDependencies` must represent the full set (not just "what we don't have yet") so that we notice dependency removals. It also pushes more change management out of the analyzer core, which should make the code easier to understand.

Reply:
Do these accumulate separately for each revision?

I presumed that pushing it twice (inside the same invocation to `preprocess(...)`) would duplicate it. I'll try and get across the accumulators a little more.

Reply:
They accumulate separately for each revision, yes, and pushing multiple times will produce duplicates. One needs to turn the vector of dependencies into a set for further processing. What I'm worried about here is that we won't catch a dependency edge between files if the file of the dependency is already in `fs` by the time `preprocess` would discover it. Then you wouldn't push to `IncludedDependencies` here.

Am I making sense? I could be misunderstanding the Salsa model.

Reply:
I think I understand how the accumulators work now. I've now changed the code so that it pushes to the `IncludedDependencies` at the beginning of that lambda.